Eco-build-Keilc/eco-dev/FFT_TEST_fft/main.temp.3
; ./main.src generated from: .\main.c
; COMPILER INVOKED BY:
;        C:\Keil\C51\BIN\C51.exe .\main.c BROWSE DEBUG CODE LISTINCLUDE SYMBOLS PREPRINT ROM(LARGE) INCDIR(../../Include) SRC(./main.src)

$NOMOD51

NAME	MAIN

PWR_UP	BIT	0A0H.7
P0	DATA	080H
P1	DATA	090H
AC	BIT	0D0H.6
T0	BIT	080H.5
SPI_CTRL	DATA	0B3H
T1	BIT	080H.6
EA	BIT	0A8H.7
T2	BIT	090H.0
CE	BIT	0A0H.6
DEV_OFFSET	DATA	0BEH
CLK1	BIT	0A0H.1
DIN0	BIT	090H.2
CLK2	BIT	0A0H.5
DIO0	BIT	090H.0
DPH0	DATA	083H
DIO1	BIT	090H.1
IE	DATA	0A8H
DPH1	DATA	085H
DIO2	BIT	080H.0
DIO3	BIT	080H.1
P0_0	BIT	080H.0
P1_0	BIT	090H.0
DIO4	BIT	080H.2
P0_1	BIT	080H.1
DPL0	DATA	082H
P1_1	BIT	090H.1
DIO5	BIT	080H.3
P0_2	BIT	080H.2
DPL1	DATA	084H
P1_2	BIT	090H.2
DIO6	BIT	080H.4
P0_3	BIT	080H.3
DIO7	BIT	080H.5
P0_4	BIT	080H.4
ADCSTATIC	DATA	0A4H
DIO8	BIT	080H.6
P0_5	BIT	080H.5
EXF2	BIT	0C8H.6
CS	BIT	0A0H.3
DIO9	BIT	080H.7
P0_6	BIT	080H.6
P0_7	BIT	080H.7
ES	BIT	0A8H.4
IP	DATA	0B8H
DATA?	BIT	0A0H.0
RI	BIT	098H.0
CY	BIT	0D0H.7
TI	BIT	098H.1
T1_1V2	DATA	0BCH
SPC_FNC	DATA	08FH
T2_1V2	DATA	0BDH
RCAP2H	DATA	0CBH
PS	BIT	0B8H.4
SP	DATA	081H
OV	BIT	0D0H.2
CK_CTRL	DATA	0B6H
RCAP2L	DATA	0CAH
TICK_DV	DATA	0B5H
RSTREAS	DATA	0B1H
C_T2	BIT	0C8H.1
EWDI	BIT	0E8H.4
ADCCON	DATA	0A1H
RCLK	BIT	0C8H.5
ADCDATAH	DATA	0A2H
EXIF	DATA	091H
TCLK	BIT	0C8H.4
DR2_CE	BIT	0A0H.6
ADCDATAL	DATA	0A3H
SBUF	DATA	099H
PCON	DATA	087H
SCON	DATA	098H
PWDI	BIT	0F8H.4
TMOD	DATA	089H
TCON	DATA	088H
WDTI	BIT	0D8H.3
PWMDUTY	DATA	0AAH
IE0	BIT	088H.1
IE1	BIT	088H.3
P0_DIR	DATA	094H
B	DATA	0F0H
P1_DIR	DATA	096H
P0_ALT	DATA	095H
P1_ALT	DATA	097H
CP_RL2	BIT	0C8H.0
DR1	BIT	0A0H.2
ACC	DATA	0E0H
DR2	BIT	0A0H.6
TEST_MODE	DATA	0B7H
SPICLK	DATA	0B4H
CT2	BIT	0C8H.1
ET0	BIT	0A8H.1
ET1	BIT	0A8H.3
TF0	BIT	088H.5
INT0_N	BIT	080H.3
ET2	BIT	0A8H.5
TF1	BIT	088H.7
INT1_N	BIT	080H.4
TF2	BIT	0C8H.7
RB8	BIT	098H.2
TH0	DATA	08CH
EX0	BIT	0A8H.0
IT0	BIT	088H.0
TH1	DATA	08DH
EX1	BIT	0A8H.2
TB8	BIT	098H.3
IT1	BIT	088H.2
TH2	DATA	0CDH
REGX_CTRL	DATA	0ADH
EX2	BIT	0E8H.0
P	BIT	0D0H.0
EX3	BIT	0E8H.1
SM0	BIT	098H.7
TL0	DATA	08AH
EX4	BIT	0E8H.2
SM1	BIT	098H.6
TL1	DATA	08BH
EX5	BIT	0E8H.3
SM2	BIT	098H.5
TL2	DATA	0CCH
EIE	DATA	0E8H
PT0	BIT	0B8H.1
RS0	BIT	0D0H.3
PT1	BIT	0B8H.3
RS1	BIT	0D0H.4
PT2	BIT	0B8H.5
TR0	BIT	088H.4
PWMCON	DATA	0A9H
TR1	BIT	088H.6
TR2	BIT	0C8H.2
PX0	BIT	0B8H.0
PX1	BIT	0B8H.2
PX2	BIT	0F8H.0
PX3	BIT	0F8H.1
PX4	BIT	0F8H.2
DPH	DATA	083H
PX5	BIT	0F8H.3
EIP	DATA	0F8H
DPL	DATA	082H
EXEN2	BIT	0C8H.3
CPRL2	BIT	0C8H.0
REN	BIT	098H.4
T2CON	DATA	0C8H
DPS	DATA	086H
SPI_DATA	DATA	0B2H
MPAGE	DATA	092H
RXD	BIT	080H.1
DOUT2	BIT	0A0H.4
EICON	DATA	0D8H
CKCON	DATA	08EH
TXD	BIT	080H.2
RADIO	DATA	0A0H
PWM	BIT	080H.7
F0	BIT	0D0H.5
F1	BIT	0D0H.1
REGX_LSB	DATA	0ACH
PSW	DATA	0D0H
REGX_MSB	DATA	0ABH
?PR?_bitrev?MAIN     SEGMENT CODE 
?DT?_bitrev?MAIN     SEGMENT DATA OVERLAYABLE 
?PR?_log_2?MAIN      SEGMENT CODE 
?PR?_compute_W?MAIN  SEGMENT CODE 
?DT?_compute_W?MAIN  SEGMENT DATA OVERLAYABLE 
?PR?_permute_bitrev?MAIN                 SEGMENT CODE 
?DT?_permute_bitrev?MAIN                 SEGMENT DATA OVERLAYABLE 
?PR?_fft?MAIN        SEGMENT CODE 
?DT?_fft?MAIN        SEGMENT DATA OVERLAYABLE 
?PR?_fft_2?MAIN        SEGMENT CODE 
?PR?_fft_3?MAIN        SEGMENT CODE 
?PR?main?MAIN        SEGMENT CODE 
?ID?_bitrev?MAIN     SEGMENT IDATA OVERLAYABLE 
?ID?_log_2?MAIN      SEGMENT IDATA OVERLAYABLE 
?ID?_compute_W?MAIN  SEGMENT IDATA OVERLAYABLE 
?ID?_permute_bitrev?MAIN                 SEGMENT IDATA OVERLAYABLE 
?ID?_fft?MAIN        SEGMENT IDATA OVERLAYABLE 
?ID?main?MAIN        SEGMENT IDATA OVERLAYABLE 
	EXTRN	CODE (_mdelay)
	EXTRN	CODE (_cos)
	EXTRN	CODE (_sin)
	EXTRN	CODE (_store_cpu_rate)
	EXTRN	CODE (?C?SIDIV)
	EXTRN	CODE (?C?IMUL)
	EXTRN	CODE (?C?FCASTI)
	EXTRN	CODE (?C?CASTF)
	EXTRN	CODE (?C?ISTOPTR)
	EXTRN	CODE (?C?ILDOPTR)
	EXTRN	CODE (?C_STARTUP)

	EXTRN   CODE (eco_page_manager)
    EXTRN   DATA (ECO_PAGE_ADDR)
    EXTRN   DATA (ECO_PAGE_SPI_CONN)
    EXTRN   IDATA (ECO_PAGE_REGISTER1)
    EXTRN   IDATA (ECO_PAGE_REGISTER2)
    EXTRN   IDATA (ECO_PAGE_REGISTER3)
    EXTRN   IDATA (ECO_PAGE_REGISTER4)
    EXTRN   IDATA (ECO_PAGE_REGISTER5)
    EXTRN   IDATA (ECO_PAGE_REGISTER6)
    EXTRN   IDATA (ECO_PAGE_REGISTER7)

	PUBLIC	main
	PUBLIC	?_fft?BYTE
	PUBLIC	_fft
	PUBLIC	_fft_2
	PUBLIC	_fft_3
	PUBLIC	?_permute_bitrev?BYTE
	PUBLIC	_permute_bitrev
	PUBLIC	?_compute_W?BYTE
	PUBLIC	_compute_W
	PUBLIC	_log_2
	PUBLIC	_bitrev

	RSEG  ?DT?_bitrev?MAIN
?_bitrev?BYTE:
    numbits?041:   DS   2

	RSEG  ?DT?_compute_W?MAIN
?_compute_W?BYTE:
          n?246:   DS   2
	ORG  2
       W_re?247:   DS   3
	ORG  5
       W_im?248:   DS   3
	ORG  8
      log2n?251:   DS   2

	RSEG  ?DT?_fft?MAIN
?_fft?BYTE:
          n?460:   DS   2
	ORG  2
       A_re?461:   DS   3
	ORG  5
       A_im?462:   DS   3
	ORG  8
       W_re?463:   DS   3
	ORG  11
       W_im?464:   DS   3

	RSEG  ?DT?_permute_bitrev?MAIN
?_permute_bitrev?BYTE:
          n?352:   DS   2
	ORG  2
       A_re?353:   DS   3
	ORG  5
       A_im?354:   DS   3

	RSEG  ?ID?_bitrev?MAIN
          i?042:   DS   2

	RSEG  ?ID?_bitrev?MAIN
        rev?043:   DS   2

	RSEG  ?ID?_log_2?MAIN
        res?145:   DS   2

	RSEG  ?ID?_compute_W?MAIN
          i?249:   DS   2

	RSEG  ?ID?_compute_W?MAIN
         br?250:   DS   2

	RSEG  ?ID?_permute_bitrev?MAIN
          i?355:   DS   2

	RSEG  ?ID?_permute_bitrev?MAIN
        bri?356:   DS   2

	RSEG  ?ID?_permute_bitrev?MAIN
      log2n?357:   DS   2

	RSEG  ?ID?_permute_bitrev?MAIN
       t_re?358:   DS   2

	RSEG  ?ID?_permute_bitrev?MAIN
       t_im?359:   DS   2

	RSEG  ?ID?_fft?MAIN
       w_re?465:   DS   2

	RSEG  ?ID?_fft?MAIN
       w_im?466:   DS   2

	RSEG  ?ID?_fft?MAIN
       u_re?467:   DS   2

	RSEG  ?ID?_fft?MAIN
       u_im?468:   DS   2

	RSEG  ?ID?_fft?MAIN
       t_re?469:   DS   2

	RSEG  ?ID?_fft?MAIN
       t_im?470:   DS   2

	RSEG  ?ID?_fft?MAIN
          m?471:   DS   2

	RSEG  ?ID?_fft?MAIN
          g?472:   DS   2

	RSEG  ?ID?_fft?MAIN
          b?473:   DS   2

	RSEG  ?ID?_fft?MAIN
         mt?474:   DS   2

	RSEG  ?ID?_fft?MAIN
          k?475:   DS   2

	RSEG  ?ID?main?MAIN
          n?576:   DS   2

	RSEG  ?ID?main?MAIN
       A_re?577:   DS   16

	RSEG  ?ID?main?MAIN
       A_im?578:   DS   16

	RSEG  ?ID?main?MAIN
       W_re?579:   DS   8

	RSEG  ?ID?main?MAIN
       W_im?580:   DS   8
; /* gets no. of points from the user, initialize the points and roots of unity lookup table 
;  * and lets fft go. finally bit-reverses the results and outputs them into a file. 
;  * n should be a power of 2. 
;  */
; #include "Eco/reg24e1.h"
; #include "Eco/eco_sys.h"
; #include "utils/utils.h"
; #include "serial/serial.h"
; #include "isr/isr_rf.h"
; #include "eeprom/eeprom.h"
; #include "eco_page.h"
; #include "adc/adc.h"
; 
; #include <math.h>
; 
; /* treats inp as a numbits number and bitreverses it. 
;  * inp < 2^(numbits) for meaningful bit-reversal
;  */ 
; int bitrev(int idata inp, int idata numbits)

	RSEG  ?PR?_bitrev?MAIN
_bitrev:
	USING	0
			; SOURCE LINE # 19
	MOV  	numbits?041,R4
	MOV  	numbits?041+01H,R5
;---- Variable 'inp?040' assigned to Register 'R2/R3' ----
	MOV  	R3,AR7
	MOV  	R2,AR6
; {
			; SOURCE LINE # 20
; 	int idata i;
; 	int idata rev=0;
			; SOURCE LINE # 22
	CLR  	A
	MOV  	R0,#LOW (rev?043)
	MOV  	@R0,A
	INC  	R0
	MOV  	@R0,A
; 	for (i=0; i < numbits; i++)
			; SOURCE LINE # 23
	MOV  	R0,#LOW (i?042)
	MOV  	@R0,A
	INC  	R0
	MOV  	@R0,A
?C0001:
	CLR  	C
	MOV  	R0,#LOW (i?042+01H)
	MOV  	A,@R0
	SUBB 	A,numbits?041+01H
	MOV  	A,numbits?041
	XRL  	A,#080H
	MOV  	B,A
	DEC  	R0
	MOV  	A,@R0
	XRL  	A,#080H
	SUBB 	A,B
	JNC  	?C0002
; 	{
			; SOURCE LINE # 24
; 		rev = (rev << 1) | (inp & 1);
			; SOURCE LINE # 25
	MOV  	R0,#LOW (rev?043+01H)
	MOV  	A,@R0
	ADD  	A,ACC
	MOV  	R7,A
	DEC  	R0
	MOV  	A,@R0
	RLC  	A
	MOV  	R6,A
	MOV  	A,R3
	ANL  	A,#01H
	MOV  	R5,A
	MOV  	A,R6
	MOV  	@R0,A
	MOV  	A,R7
	ORL  	A,R5
	INC  	R0
	MOV  	@R0,A
; 		inp >>= 1;
			; SOURCE LINE # 26
	MOV  	A,R2
	MOV  	C,ACC.7
	RRC  	A
	MOV  	R2,A
	MOV  	A,R3
	RRC  	A
	MOV  	R3,A
; 	}
			; SOURCE LINE # 27
	MOV  	R0,#LOW (i?042+01H)
	INC  	@R0
	MOV  	A,@R0
	DEC  	R0
	JNZ  	?C0001
	INC  	@R0
?C0034:
	SJMP 	?C0001
?C0002:
; 	return rev;
			; SOURCE LINE # 28
	MOV  	R0,#LOW (rev?043)
	MOV  	A,@R0
	MOV  	R6,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R7,A
; }
			; SOURCE LINE # 29
?C0004:
	RET  	
; END OF _bitrev

; 
; 
; /* returns log n (to the base 2), if n is positive and power of 2 */ 
; int log_2(int idata n) 

	RSEG  ?PR?_log_2?MAIN
_log_2:
	USING	0
			; SOURCE LINE # 33
;---- Variable 'n?144' assigned to Register 'R6/R7' ----
; {
			; SOURCE LINE # 34
; 	int idata res; 
; 	for (res=0; n >= 2; res++) 
			; SOURCE LINE # 36
	CLR  	A
	MOV  	R0,#LOW (res?145)
	MOV  	@R0,A
	INC  	R0
	MOV  	@R0,A
?C0005:
	CLR  	C
	MOV  	A,R7
	SUBB 	A,#02H
	MOV  	A,R6
	XRL  	A,#080H
	SUBB 	A,#080H
	JC   	?C0006
; 		n = n >> 1; 
			; SOURCE LINE # 37
	MOV  	A,R6
	MOV  	C,ACC.7
	RRC  	A
	MOV  	R6,A
	MOV  	A,R7
	RRC  	A
	MOV  	R7,A
	MOV  	R0,#LOW (res?145+01H)
	INC  	@R0
	MOV  	A,@R0
	DEC  	R0
	JNZ  	?C0005
	INC  	@R0
?C0035:
	SJMP 	?C0005
?C0006:
; 	return res; 
			; SOURCE LINE # 38
	MOV  	R0,#LOW (res?145)
	MOV  	A,@R0
	MOV  	R6,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R7,A
; }
			; SOURCE LINE # 39
?C0008:
	RET  	
; END OF _log_2

; 
; 
; 
; /* W will contain roots of unity so that W[bitrev(i,log2n-1)] = e^(2*pi*i/n)
;  * n should be a power of 2
;  * Note: W is bit-reversal permuted because fft(..) goes faster if this is done.
;  *       see that function for more details on why we treat 'i' as a (log2n-1) bit number.
;  */
; void compute_W(int idata n, int idata W_re[], int idata W_im[])

	RSEG  ?PR?_compute_W?MAIN
_compute_W:
	USING	0
			; SOURCE LINE # 48
	MOV  	n?246,R6
	MOV  	n?246+01H,R7
	MOV  	W_re?247,R3
	MOV  	W_re?247+01H,R2
	MOV  	W_re?247+02H,R1
; {
			; SOURCE LINE # 49
; 	int idata i;
; 	int idata br;
; 	int log2n = log_2(n);
			; SOURCE LINE # 52
	LCALL	_log_2
	MOV  	log2n?251,R6
	MOV  	log2n?251+01H,R7
; 
; 	for (i=0; i<(n/2); i++)
			; SOURCE LINE # 54
	CLR  	A
	MOV  	R0,#LOW (i?249)
	MOV  	@R0,A
	INC  	R0
	MOV  	@R0,A
?C0009:
	MOV  	R6,n?246
	MOV  	R7,n?246+01H
	MOV  	R4,#00H
	MOV  	R5,#02H
	LCALL	?C?SIDIV
	MOV  	R4,AR6
	MOV  	R5,AR7
	MOV  	R0,#LOW (i?249)
	MOV  	A,@R0
	MOV  	R6,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R7,A
	CLR  	C
	SUBB 	A,R5
	MOV  	A,R4
	XRL  	A,#080H
	MOV  	R0,A
	MOV  	A,R6
	XRL  	A,#080H
	SUBB 	A,R0
	JC   	$ + 5H
	LJMP 	?C0012
; 	{
			; SOURCE LINE # 55
; 		br = bitrev(i,log2n-1); 
			; SOURCE LINE # 56
	MOV  	A,log2n?251+01H
	ADD  	A,#0FFH
	MOV  	R5,A
	MOV  	A,log2n?251
	ADDC 	A,#0FFH
	MOV  	R4,A
	LCALL	_bitrev
	MOV  	R0,#LOW (br?250)
	MOV  	@R0,AR6
	INC  	R0
	MOV  	@R0,AR7
; 		W_re[br] = cos((i*2*3)/(n));  
			; SOURCE LINE # 57
	MOV  	R0,#LOW (i?249)
	MOV  	A,@R0
	MOV  	R6,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R7,A
	MOV  	R4,#00H
	MOV  	R5,#06H
	LCALL	?C?IMUL
	MOV  	R4,n?246
	MOV  	R5,n?246+01H
	LCALL	?C?SIDIV
	MOV  	R4,AR6
	MOV  	R5,AR7
	MOV  	A,R4
	LCALL	?C?FCASTI
	LCALL	_cos
	LCALL	?C?CASTF
	MOV  	R5,AR7
	MOV  	R0,#LOW (br?250+01H)
	MOV  	A,@R0
	ADD  	A,ACC
	MOV  	R7,A
	DEC  	R0
	MOV  	A,@R0
	RLC  	A
	MOV  	R3,W_re?247
	MOV  	R2,W_re?247+01H
	MOV  	R1,W_re?247+02H
	MOV  	DPL,R7
	MOV  	DPH,A
	MOV  	A,R6
	MOV  	B,R5
	LCALL	?C?ISTOPTR
; 		W_im[br] = sin((i*2*3)/(n));  
			; SOURCE LINE # 58
	MOV  	R0,#LOW (i?249)
	MOV  	A,@R0
	MOV  	R6,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R7,A
	MOV  	R4,#00H
	MOV  	R5,#06H
	LCALL	?C?IMUL
	MOV  	R4,n?246
	MOV  	R5,n?246+01H
	LCALL	?C?SIDIV
	MOV  	R4,AR6
	MOV  	R5,AR7
	MOV  	A,R4
	LCALL	?C?FCASTI
	LCALL	_sin
	LCALL	?C?CASTF
	MOV  	R5,AR7
	MOV  	R0,#LOW (br?250+01H)
	MOV  	A,@R0
	ADD  	A,ACC
	MOV  	R7,A
	DEC  	R0
	MOV  	A,@R0
	RLC  	A
	MOV  	R3,W_im?248
	MOV  	R2,W_im?248+01H
	MOV  	R1,W_im?248+02H
	MOV  	DPL,R7
	MOV  	DPH,A
	MOV  	A,R6
	MOV  	B,R5
	LCALL	?C?ISTOPTR
; 	}
			; SOURCE LINE # 59
	MOV  	R0,#LOW (i?249+01H)
	INC  	@R0
	MOV  	A,@R0
	DEC  	R0
	JZ   	$ + 5H
	LJMP 	?C0009
	INC  	@R0
?C0036:
	LJMP 	?C0009
; //#ifdef COMMENT_ONLY 
; //	for (i=0;i<(n/2);i++)
; //	{ 
; //		br = i; //bitrev(i,log2n-1); 
; //		printf("(%g\t%g)\n", W_re[br], W_im[br]);
; //	}  
; //#endif 
; }
			; SOURCE LINE # 67
?C0012:
	RET  	
; END OF _compute_W

; 
; 
; /* permutes the array using a bit-reversal permutation */ 
; void permute_bitrev(int n, int *A_re, int *A_im) 

	RSEG  ?PR?_permute_bitrev?MAIN
_permute_bitrev:
	USING	0
			; SOURCE LINE # 71
	MOV  	n?352,R6
	MOV  	n?352+01H,R7
	MOV  	A_re?353,R3
	MOV  	A_re?353+01H,R2
	MOV  	A_re?353+02H,R1
; { 
			; SOURCE LINE # 72
; 	int idata i;
; 	int idata bri;
; 	int idata log2n;
; 	int idata t_re;
; 	int idata t_im;
; 
; 	log2n = log_2(n); 
			; SOURCE LINE # 79
	LCALL	_log_2
	MOV  	R0,#LOW (log2n?357)
	MOV  	@R0,AR6
	INC  	R0
	MOV  	@R0,AR7
; 
; 	for (i=0; i<n; i++)
			; SOURCE LINE # 81
	CLR  	A
	MOV  	R0,#LOW (i?355)
	MOV  	@R0,A
	INC  	R0
	MOV  	@R0,A
?C0013:
	MOV  	R0,#LOW (i?355)
	MOV  	A,@R0
	MOV  	R6,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R7,A
	CLR  	C
	SUBB 	A,n?352+01H
	MOV  	A,n?352
	XRL  	A,#080H
	MOV  	R0,A
	MOV  	A,R6
	XRL  	A,#080H
	SUBB 	A,R0
	JC   	$ + 5H
	LJMP 	?C0017
; 	{
			; SOURCE LINE # 82
; 		bri  = bitrev(i, log2n);
			; SOURCE LINE # 83
	MOV  	R0,#LOW (log2n?357)
	MOV  	A,@R0
	MOV  	R4,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R5,A
	LCALL	_bitrev
	MOV  	R0,#LOW (bri?356)
	MOV  	A,R6
	MOV  	@R0,A
	INC  	R0
	MOV  	A,R7
	MOV  	@R0,A
; 
; 		/* skip already swapped elements */
; 		if (bri <= i) continue;
			; SOURCE LINE # 86
	MOV  	R0,#LOW (i?355)
	MOV  	A,@R0
	MOV  	R4,A
	INC  	R0
	MOV  	A,@R0
	MOV  	R5,A
	SETB 	C
	MOV  	A,R7
	SUBB 	A,R5
	MOV  	A,R4
	XRL  	A,#080H
	MOV  	R0,A
	MOV  	A,R6
	XRL  	A,#080H
	SUBB 	A,R0
	JNC  	$ + 5H
	LJMP 	?C0015
?C0016:
; 
; 		t_re = A_re[i];
			; SOURCE LINE # 88
	MOV  	R0,#LOW (i?355+01H)
	MOV  	A,@R0
	ADD  	A,ACC
	MOV  	R7,A
	DEC  	R0
	MOV  	A,@R0
	RLC  	A
	MOV  	R6,A

Eco-build-Keilc/eco-dev/sha/main.c
#include "Eco/reg24e1.h"
#include "Eco/eco_sys.h"
#include "utils/utils.h"
#include "serial/serial.h"
//#include "isr/isr_rf.h"
#include "eeprom/eeprom.h"
#include "eco_page.h"
//#include "adc/adc.h"

#define TESTA   "abc"
#define SHA1CircularShift(bits,word) \
                ((((word) << (bits)) & 0xFFFFFFFF) | \
                ((word) >> (32-(bits))))


typedef struct SHA1Context
{
    unsigned Message_Digest[5]; /* Message Digest (output)          */

    unsigned Length_Low;        /* Message length in bits           */
    unsigned Length_High;       /* Message length in bits           */

    unsigned char Message_Block[64]; /* 256-bit message blocks      */
    char Message_Block_Index;    /* Index into message block array   */

    int Computed;               /* Is the digest computed?          */
    int Corrupted;              /* Is the message digest corruped?  */
} SHA1Context;


void SHA1ProcessMessageBlock(SHA1Context idata *context)
{
	idata unsigned K[] =            /* Constants defined in SHA-1   */      
    {
        0x5A827999,
        0x6ED9EBA1,
        0x8F1BBCDC,
        0xCA62C1D6
    };
    idata int         t;                  /* Loop counter                 */
    idata unsigned    temp;               /* Temporary word value         */
    xdata unsigned    W[80];              /* Word sequence                */
    idata unsigned    A, B, C, D, E;      /* Word buffers                 */

    /*
     *  Initialize the first 16 words in the array W
     */
    for(t = 0; t < 16; t++)
    {
        W[t] = ((unsigned) context->Message_Block[t * 4]) << 24;
        W[t] |= ((unsigned) context->Message_Block[t * 4 + 1]) << 16;
        W[t] |= ((unsigned) context->Message_Block[t * 4 + 2]) << 8;
        W[t] |= ((unsigned) context->Message_Block[t * 4 + 3]);
    }

    for(t = 16; t < 80; t++)
    {
       W[t] = SHA1CircularShift(1,W[t-3] ^ W[t-8] ^ W[t-14] ^ W[t-16]);
    }

    A = context->Message_Digest[0];
    B = context->Message_Digest[1];
    C = context->Message_Digest[2];
    D = context->Message_Digest[3];
    E = context->Message_Digest[4];

    for(t = 0; t < 20; t++)
    {
        temp =  SHA1CircularShift(5,A) +
                ((B & C) | ((~B) & D)) + E + W[t] + K[0];
        temp &= 0xFFFFFFFF;
        E = D;
        D = C;
        C = SHA1CircularShift(30,B);
        B = A;
        A = temp;
    }

    for(t = 20; t < 40; t++)
    {
        temp = SHA1CircularShift(5,A) + (B ^ C ^ D) + E + W[t] + K[1];
        temp &= 0xFFFFFFFF;
        E = D;
        D = C;
        C = SHA1CircularShift(30,B);
        B = A;
        A = temp;
    }

    for(t = 40; t < 60; t++)
    {
        temp = SHA1CircularShift(5,A) +
               ((B & C) | (B & D) | (C & D)) + E + W[t] + K[2];
        temp &= 0xFFFFFFFF;
        E = D;
        D = C;
        C = SHA1CircularShift(30,B);
        B = A;
        A = temp;
    }

    for(t = 60; t < 80; t++)
    {
        temp = SHA1CircularShift(5,A) + (B ^ C ^ D) + E + W[t] + K[3];
        temp &= 0xFFFFFFFF;
        E = D;
        D = C;
        C = SHA1CircularShift(30,B);
        B = A;
        A = temp;
    }

    context->Message_Digest[0] =
                        (context->Message_Digest[0] + A) & 0xFFFFFFFF;
    context->Message_Digest[1] =
                        (context->Message_Digest[1] + B) & 0xFFFFFFFF;
    context->Message_Digest[2] =
                        (context->Message_Digest[2] + C) & 0xFFFFFFFF;
    context->Message_Digest[3] =
                        (context->Message_Digest[3] + D) & 0xFFFFFFFF;
    context->Message_Digest[4] =
                        (context->Message_Digest[4] + E) & 0xFFFFFFFF;

    context->Message_Block_Index = 0;

}


void SHA1Reset(SHA1Context idata *context)
{
    context->Length_Low             = 0;
    context->Length_High            = 0;
    context->Message_Block_Index    = 0;

    context->Message_Digest[0]      = 0x67452301;
    context->Message_Digest[1]      = 0xEFCDAB89;
    context->Message_Digest[2]      = 0x98BADCFE;
    context->Message_Digest[3]      = 0x10325476;
    context->Message_Digest[4]      = 0xC3D2E1F0;

    context->Computed   = 0;
    context->Corrupted  = 0;
}


void SHA1PadMessage(SHA1Context idata *context)
{
    /*
     *  Check to see if the current message block is too small to hold
     *  the initial padding bits and length.  If so, we will pad the
     *  block, process it, and then continue padding into a second
     *  block.
     */
    if (context->Message_Block_Index > 55)
    {
        context->Message_Block[context->Message_Block_Index++] = 0x80;
        while(context->Message_Block_Index < 64)
        {
            context->Message_Block[context->Message_Block_Index++] = 0;
        }

        SHA1ProcessMessageBlock(context);

        while(context->Message_Block_Index < 56)
        {
            context->Message_Block[context->Message_Block_Index++] = 0;
        }
    }
    else
    {
        context->Message_Block[context->Message_Block_Index++] = 0x80;
        while(context->Message_Block_Index < 56)
        {
            context->Message_Block[context->Message_Block_Index++] = 0;
        }
    }

    /*
     *  Store the message length as the last 8 octets
     */
    context->Message_Block[24] = (context->Length_High >> 24) & 0xFF;
    context->Message_Block[25] = (context->Length_High >> 16) & 0xFF;
    context->Message_Block[26] = (context->Length_High >> 8) & 0xFF;
    context->Message_Block[27] = (context->Length_High) & 0xFF;
    context->Message_Block[28] = (context->Length_Low >> 24) & 0xFF;
    context->Message_Block[29] = (context->Length_Low >> 16) & 0xFF;
    context->Message_Block[30] = (context->Length_Low >> 8) & 0xFF;
    context->Message_Block[31] = (context->Length_Low) & 0xFF;

    SHA1ProcessMessageBlock(context);
}


int SHA1Result(SHA1Context idata *context)
{

    if (context->Corrupted)
    {
        return 0;
    }

    if (!context->Computed)
    {
        SHA1PadMessage(context);
        context->Computed = 1;
    }

    return 1;
}


void SHA1Input(SHA1Context idata *context,unsigned char idata *message_array,unsigned idata length)
{
    if (!length)
    {
        return;
    }

    if (context->Computed || context->Corrupted)
    {
        context->Corrupted = 1;
        return;
    }

    while(length-- && !context->Corrupted)
    {
        context->Message_Block[context->Message_Block_Index++] =(*message_array & 0xFF);

        context->Length_Low += 8;
        /* Force it to 32 bits */
        context->Length_Low &= 0xFFFFFFFF;
        if (context->Length_Low == 0)
        {
            context->Length_High++;
            /* Force it to 32 bits */
            context->Length_High &= 0xFFFFFFFF;
            if (context->Length_High == 0)
            {
                /* Message is too long */
                context->Corrupted = 1;
            }
        }

        if (context->Message_Block_Index == 32)
        {
            SHA1ProcessMessageBlock(context);
        }

        message_array++;
    }
}


void main()
{

	idata char i;
	idata SHA1Context sha;

	store_cpu_rate(16);

    P0_DIR &= ~0x28;
    P0_ALT &= ~0x28;

	serial_init(19200);


	for(i=0;i<6;i++)
	{
		blink_led();
		mdelay(400);
	}	

	while(1)
	{
		SHA1Reset(&sha);
		SHA1Input(&sha, (const unsigned char *) TESTA, 3);

		SHA1Result(&sha);
		blink_led();	
	}	

}


